### PR TITLE
[API-1250] Move type name and class to the `CompactSerializer`

### PR DESCRIPTION
**Pull request by mdumandag**
This PR updates the `CompactSerializer` interface, by moving
the type name and class definitions to there.

The newly added method names to the `CompactSerializer` API
are `getTypeName` and `getCompactClass` (because `getClass` is a method
defined in the `Object` class).

Due to that, the APIs in the `CompactSerializationConfig` are
also changed.

We have decided to use the `addSerializer` API to register
serializers, and `addClass` API to register classes to be serialized
with Compact serialization, overriding other serialization mechanisms.

Also, the declarative configuration is also updated to reflect
these changes, it now looks like

```xml
<compact-serialization enabled="true">
	<serializers>
		<serializer>com.example.FooSerializer</serializer>
	</serializers>
	<classes>
		<class>com.example.Foo</class>
	</classes>
</compact-serialization>
```

```yaml
compact-serialization:
  enabled: true
  serializers:
    - serializer: "com.example.FooSerializer"
  classes:
    - class: "com.example.Foo"
```

The motivation is to make the API consistent across all clients.

### CODE REVIEW

**Comment by mdumandag** [comment]
run-ee-compile

File: hazelcast/src/main/java/com/hazelcast/config/CompactSerializationConfig.java
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/config/CompactSerializationConfig.java:63]: Will we change this since tag everywhere with :
```suggestion
 * @since Hazelcast 5.2
```

File: hazelcast/src/main/java/com/hazelcast/internal/config/ConfigXmlGeneratorHelper.java
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/internal/config/ConfigXmlGeneratorHelper.java:50]: why don't we remove this in 5.2-BETA?

File: hazelcast/src/main/resources/hazelcast-client-config-5.2.xsd
Comment (ihsandemir) [inline hazelcast/src/main/resources/hazelcast-client-config-5.2.xsd:707]: why do we have `compact-` prefix?

**Review by ihsandemir — COMMENTED** [review]
Do you have a test which uses example xml and yaml configs for the compact classes and serializers configs?

File: hazelcast-spring/src/main/resources/hazelcast-spring-5.2.xsd
Comment (mdumandag) [reply hazelcast-spring/src/main/resources/hazelcast-spring-5.2.xsd:1538]: This is just a complex type definition, which is used in `compact-serialization`. 

It is not named as `serializers` as the type is defined separately, to avoid possible confusion.

Note that, the name of the complex type has nothing to do with how it is used. The user provides it as `<serializers></serializers>` regardless of its complex type name

File: hazelcast/src/main/java/com/hazelcast/config/CompactSerializationConfig.java
Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/config/CompactSerializationConfig.java:63]: Yeah, I was waiting for all the PRs to be merged. There are lots of places where we have such beta tags and since values

File: hazelcast/src/main/java/com/hazelcast/internal/config/ConfigXmlGeneratorHelper.java
Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/internal/config/ConfigXmlGeneratorHelper.java:50]: because we haven't merged all the PRs. once we do it for 5.2-BETA, I will remove this, along with other places

File: hazelcast/src/main/resources/hazelcast-client-config-5.2.xsd
Comment (mdumandag) [reply hazelcast/src/main/resources/hazelcast-client-config-5.2.xsd:707]: See my comment above, this is like this to avoid possible confusion, as the type definition is not under the compact serializer tag

**Comment by mdumandag** [comment]
@ihsandemir yes we do have lots of tests that validate those XML and YAML configs. They are in this PR

**Comment by mdumandag** [comment]
@ihsandemir I have updated the method name as you suggested

File: hazelcast/src/main/java/com/hazelcast/internal/config/ConfigXmlGeneratorHelper.java
Comment (ihsandemir) [reply hazelcast/src/main/java/com/hazelcast/internal/config/ConfigXmlGeneratorHelper.java:50]: will you do it for 5.2-BETA release? I think we should.

**Review by ihsandemir — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/internal/config/ConfigXmlGeneratorHelper.java
Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/internal/config/ConfigXmlGeneratorHelper.java:50]: Yes, we should. I was planning to do it after merging these PRs + schema distribution PR

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java
Comment (srknzl) [inline hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:239]: You said multiple serializer registrations in the test, but do only a single serializer registration here.

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java
Comment (srknzl) [inline hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:254]: Similar here, you are not doing multiple class registrations here.

**Review by srknzl — CHANGES_REQUESTED** [review]

File: hazelcast-spring/src/main/java/com/hazelcast/spring/AbstractHazelcastBeanDefinitionParser.java
Comment (mdumandag) [reply hazelcast-spring/src/main/java/com/hazelcast/spring/AbstractHazelcastBeanDefinitionParser.java:380]: We will catch it while building the serialization service and fail-fast with an error message explaining the problem

File: hazelcast/src/main/java/com/hazelcast/internal/config/ConfigXmlGeneratorHelper.java
Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/internal/config/ConfigXmlGeneratorHelper.java:71]: No, you can either use programmatic or declarative configuration. This is like this to handle both of the cases.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java
Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java:292]: We can't have the checks for the declarative configuration as we need to load classes to do the checks (which can happen at the serialization service construction earliest). So, we need those checks for the declarative configuration, and I re-used the same code for the programmatic configuration (which has the checks as you described)

File: hazelcast/src/main/resources/hazelcast-client-full-example.xml
Comment (mdumandag) [reply hazelcast/src/main/resources/hazelcast-client-full-example.xml:34]: These are not really in scope of this PR, but I will update them as well

File: hazelcast/src/main/resources/hazelcast-config-5.2.xsd
Comment (mdumandag) [reply hazelcast/src/main/resources/hazelcast-config-5.2.xsd:3129]: No, the `maxOccurs` for those elements is 1.

File: hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java
Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java:745]: No, we are adding one class and one serializer here. So, we need to check both of them

File: hazelcast/src/test/java/com/hazelcast/client/config/YamlClientConfigBuilderTest.java
Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/client/config/YamlClientConfigBuilderTest.java:736]: No, we are adding one class and one serializer here. So, we need to check both of them

File: hazelcast/src/test/java/example/serialization/SameTypeNameEmployeeDTOSerializer.java
Comment (mdumandag) [reply hazelcast/src/test/java/example/serialization/SameTypeNameEmployeeDTOSerializer.java:31]: No, it is Employe**r**DTO, not Employe**e**DTO

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java
Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:239]: What I meant with multiple serializer registrations was the method (setSerializers). Let me update the method names to make it more clear

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java
Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java:254]: Same as above

File: hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java
Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java:709]: Valid point, I have updated the method to make use of a serializable object and add an assertion about it being serialized as compact

File: hazelcast/src/main/resources/hazelcast-client-full-example.xml
Comment (mdumandag) [reply hazelcast/src/main/resources/hazelcast-client-full-example.xml:34]: On a second though, I decided not to do it :smile: As I said, they are not related to this PR, and there are too many of them, which would pollute this PR a bit to much

File: hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java
Comment (srknzl) [reply hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java:709]: Can we also somehow assert that reflective serializer is not used ? I think that's what we mean by "explicit" serializer is used.

Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/client/config/XmlClientConfigBuilderTest.java:709]: Since I made it Serializable, if the explicit serializer wouldn't be used, it wouldn't be Compact, as Java serializable has more priority over reflective serializer

**Review by srknzl — APPROVED** [review]

**Pull request merged**
